### ISSUE TITLE
Deprecation error when calling `sum` on objects that can be summed

### ISSUE THREAD

**Issue by ghiculescu** [rails/rails#46779]
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
end

require "active_support"
require "active_support/all"
require "minitest/autorun"

class Money
  attr_reader :value

  def initialize(value)
    @value = value
  end

  def +(other)
    Money.new(value + other.value)
  end

 def coerce(other)
   [Money.new(other), self]
 end

  def ==(other)
    other.value == value
  end
end

class BugTest < ActiveSupport::TestCase
  def test_stuff
    assert_equal Money.new(3), [Money.new(1), Money.new(2)].sum
  end
end
```

### Expected behavior

Deprecation warning:

```
DEPRECATION WARNING: Rails 7.0 has deprecated Enumerable.sum in favor of Ruby's native implementation available since 2.4. Sum of non-numeric elements requires an initial argument. (called from test_stuff)
```

### Actual behavior

No warning.

I don't expect a warning, because the relevant code works outside a Rails context fine. Because the Money class (which is based on [this gem](https://github.com/Shopify/money)) has a `coerce` method it is already able to handle being added to other numerics, and so the native implementation works correctly even without an initial argument.

Here is proof the code works with no Rails, and no warnings:

<details>
<summary>Rails-free executable test case</summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "minitest"
end

require "minitest/autorun"

class Money
  attr_reader :value

  def initialize(value)
    @value = value
  end

  def +(other)
    Money.new(value + other.value)
  end

 def coerce(other)
   [Money.new(other), self]
 end

  def ==(other)
    other.value == value
  end
end

class BugTest < Minitest::Test
  def test_stuff
    assert_equal Money.new(3), [Money.new(1), Money.new(2)].sum # deprecation warning here
  end
end
```

</details>

The discussion in https://github.com/rails/rails/pull/41669 also discussed Money objects.

I am not sure exactly which code paths will get removed in 7.1 as a result of https://github.com/rails/rails/pull/42080 - it's possible this will get fixed as part of the deprecated code being removed.

So I guess the question is, is it worth trying to fix the logic for the deprecation warning between now and 7.1?

### System configuration
**Rails version**: `main`

**Ruby version**: 3.1.3

**Comment by yahonda** [comment]
Let me clarify what is an actual behavior because the reproduction script raises the `DEPRECATION WARNING:`.

```ruby
% ruby -v
ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [x86_64-darwin22]
% ruby rep46779.rb
Fetching https://github.com/rails/rails.git
Resolving dependencies...Fetching gem metadata from https://rubygems.org/.......
..........
Using rake 13.0.6
Using concurrent-ruby 1.1.10
Using zeitwerk 2.6.6
Using date 3.3.3
Using marcel 1.0.2
Using bundler 2.3.26
Using racc 1.6.1
Using crass 1.0.6
Using i18n 1.12.0
Using rack 2.2.4
Using websocket-extensions 0.1.5
Using connection_pool 2.3.0
Using mini_mime 1.1.2
Using minitest 5.16.3
Using builder 3.2.4
Using timeout 0.3.1
Using erubi 1.11.0
Using thor 1.2.1
Using tzinfo 2.0.5
Using nio4r 2.5.8
Using activesupport 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using rack-test 2.0.2
Using websocket-driver 0.7.5
Using net-protocol 0.2.1
Using nokogiri 1.13.10 (x86_64-darwin)
Using net-smtp 0.3.3
Using net-imap 0.3.2
Using net-pop 0.1.2
Using loofah 2.19.1
Using mail 2.8.0
Using rails-html-sanitizer 1.4.4
Using rails-dom-testing 2.0.3
Using activemodel 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using actionview 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using globalid 1.0.0
Using actionpack 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using actioncable 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using activejob 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using activerecord 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using railties 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using activestorage 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using actionmailer 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using actionmailbox 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using actiontext 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Using rails 7.1.0.alpha from https://github.com/rails/rails.git (at main@cd762a8)
Run options: --seed 10580

# Running:

DEPRECATION WARNING: Rails 7.0 has deprecated Enumerable.sum in favor of Ruby's native implementation available since 2.4. Sum of non-numeric elements requires an initial argument. (called from test_stuff at rep46779.rb:39)
.

Finished in 0.001247s, 801.9246 runs/s, 801.9246 assertions/s.
1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
%
```

```ruby
% more rep46779.rb
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
end

require "active_support"
require "active_support/all"
require "minitest/autorun"

class Money
  attr_reader :value

  def initialize(value)
    @value = value
  end

  def +(other)
    Money.new(value + other.value)
  end

 def coerce(other)
   [Money.new(other), self]
 end

  def ==(other)
    other.value == value
  end
end

class BugTest < ActiveSupport::TestCase
  def test_stuff
    assert_equal Money.new(3), [Money.new(1), Money.new(2)].sum
  end
end
%
```

**Comment by ghiculescu** [comment]
Yes that's right. But the "Rails-free executable test case" does *not* get a deprecation warning, and still works fine.

The implication of the deprecation warning is that I should change my code (I should change `.sum` to `.sum(Money.new(0))` or something like that), but that is incorrect, as the second script shows.

**Comment by nvasilevski** [comment]
That's actually a very good find. We have a bunch of `Enumerable#sum` deprecations we are yet to fix and many of them caused by exactly the "`sum` on an array of `Money` objects" case. It would be very helpful not to issue a warning for such cases. It also makes me think I should check if adding `coerce` method to other non-numeric objects would make more sense than providing an initial value to the `sum` method

**Comment by ghiculescu** [comment]
> I should check if adding `coerce` method to other non-numeric objects would make more sense than providing an initial value to the `sum` method

In terms of nicer code, I think it would. But it would result of more of these warnings :')

**Comment by yahonda** [comment]
> Yes that's right. But the "Rails-free executable test case" does not get a deprecation warning, and still works fine.

I got it. Thanks.

**Comment by ghiculescu** [comment]
That deprecation warning has been in Rails for about a year and a half now. And it has been in a release for [over a year](https://rubyonrails.org/2021/12/15/Rails-7-fulfilling-a-vision).

So given that, I think one way to fix this issue would be to just go ahead with the deprecation. Assuming this is the intended change:

```diff
diff --git a/activesupport/lib/active_support/core_ext/enumerable.rb b/activesupport/lib/active_support/core_ext/enumerable.rb
index e49ff7b256..5843a53111 100644
--- a/activesupport/lib/active_support/core_ext/enumerable.rb
+++ b/activesupport/lib/active_support/core_ext/enumerable.rb
@@ -84,11 +84,7 @@ def sum(identity = nil, &block)
       identity ||= 0
       _original_sum_with_required_identity(identity, &block)
     else
-      ActiveSupport.deprecator.warn(<<-MSG.squish)
-        Rails 7.0 has deprecated Enumerable.sum in favor of Ruby's native implementation available since 2.4.
-        Sum of non-numeric elements requires an initial argument.
-      MSG
-      inject(:+) || 0
+      super
     end
   end
```

@yahonda @rafaelfranca (tagging you since it's deprecations related), does this match your expectation for this deprecation? Would you be open to a PR for this?

**Comment by skipkayhil** [comment]
> Assuming this is the intended change:

I believe the goal of the deprecation is to stop overriding `Enumerable#sum` at all. So I'd expect the change to remove `#sum` and the `_original_sum_with_required_identity` alias above.

**Comment by ghiculescu** [comment]
I think that would also solve this issue:

There's 3 code paths in `Enumerable#sum` that aren't deprecated. The first and third call `_original_sum_with_required_identity` so that's fine. The second one, 

```ruby
    elsif block_given?
      map(&block).sum
```

seems to work the same with Ruby's `sum`, so it probably also could be calling `_original_sum_with_required_identity`, and is safe to remove.

```ruby
# irb
3.1.3 :005 > ["a", "b", "c"].sum(&:length)
 => 3
```

tl;dr yes I think removing the override would also solve this issue! When can we do that?

**Comment by jedrekdomanski** [comment]
I agree, I think the author of https://github.com/rails/rails/pull/42080 didn't take into account the case when you sum a list of `Money` objects which works outside Rails context:
```ruby
require 'money'
[Money.new(1000), Money.new(1000), Money.new(1000)].sum
=> #<Money fractional:3000 currency:USD>
```
It would be a waste of time to change all those occurrences in the entire project to just get rid of the warnings.
